CI fails with dev dependencies: RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1

CI fails with dev dependencies: https://github.com/huggingface/trl/actions/runs/28625219492/job/84890038788
> RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1

```python
  FAILED tests/test_sft_trainer.py::TestPatchChunkedCELMHead::test_backward_matches_reference[trl-internal-testing/tiny-DeepseekV3ForCausalLM] - RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  FAILED tests/test_sft_trainer.py::TestPatchChunkedCELMHead::test_backward_matches_reference[trl-internal-testing/tiny-DeepseekV3ForCausalLM-0528] - RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  FAILED tests/test_utils.py::TestPatchChunkedLMHead::test_forward[1.0-trl-internal-testing/tiny-DeepseekV3ForCausalLM] - RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  FAILED tests/test_utils.py::TestPatchChunkedLMHead::test_forward[1.0-trl-internal-testing/tiny-DeepseekV3ForCausalLM-0528] - RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  FAILED tests/test_utils.py::TestPatchChunkedLMHead::test_forward[0.7-trl-internal-testing/tiny-DeepseekV3ForCausalLM] - RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  FAILED tests/test_utils.py::TestPatchChunkedLMHead::test_forward[0.7-trl-internal-testing/tiny-DeepseekV3ForCausalLM-0528] - RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  FAILED tests/test_utils.py::TestPatchChunkedLMHead::test_backward[1.0-trl-internal-testing/tiny-DeepseekV3ForCausalLM] - RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  FAILED tests/test_utils.py::TestPatchChunkedLMHead::test_backward[1.0-trl-internal-testing/tiny-DeepseekV3ForCausalLM-0528] - RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  FAILED tests/test_utils.py::TestPatchChunkedLMHead::test_backward[0.7-trl-internal-testing/tiny-DeepseekV3ForCausalLM] - RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  FAILED tests/test_utils.py::TestPatchChunkedLMHead::test_backward[0.7-trl-internal-testing/tiny-DeepseekV3ForCausalLM-0528] - RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  FAILED tests/test_sft_trainer.py::TestPatchChunkedCELMHead::test_forward_matches_reference[trl-internal-testing/tiny-DeepseekV3ForCausalLM] - RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  FAILED tests/test_sft_trainer.py::TestPatchChunkedCELMHead::test_forward_matches_reference[trl-internal-testing/tiny-DeepseekV3ForCausalLM-0528] - RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  = 12 failed
```

Traceback:
```python
   _ TestPatchChunkedCELMHead.test_forward_matches_reference[trl-internal-testing/tiny-DeepseekV3ForCausalLM-0528] _
  [gw1] linux -- Python 3.12.13 /__w/trl/trl/.venv/bin/python3
  
  self = <tests.test_sft_trainer.TestPatchChunkedCELMHead object at 0x7f22518c18b0>
  model_id = 'trl-internal-testing/tiny-DeepseekV3ForCausalLM-0528'
  
      @pytest.mark.parametrize("model_id", _CHUNKED_CE_MODEL_IDS)
      def test_forward_matches_reference(self, model_id):
          ref_model, chunked_model, input_ids, labels, num_items = self._setup(model_id)
      
          with torch.no_grad():
  >           ref_out = ref_model(input_ids=input_ids, labels=labels, num_items_in_batch=num_items)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  tests/test_sft_trainer.py:2729: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  .venv/lib/python3.12/site-packages/torch/nn/modules/module.py:1778: in _wrapped_call_impl
      return self._call_impl(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  .venv/lib/python3.12/site-packages/torch/nn/modules/module.py:1789: in _call_impl
      return forward_call(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  .venv/lib/python3.12/site-packages/transformers/utils/generic.py:911: in wrapper
      output = func(self, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  .venv/lib/python3.12/site-packages/transformers/models/deepseek_v3/modeling_deepseek_v3.py:675: in forward
      outputs: BaseModelOutputWithPast = self.model(
  .venv/lib/python3.12/site-packages/torch/nn/modules/module.py:1778: in _wrapped_call_impl
      return self._call_impl(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  .venv/lib/python3.12/site-packages/torch/nn/modules/module.py:1789: in _call_impl
      return forward_call(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  .venv/lib/python3.12/site-packages/transformers/utils/generic.py:1040: in wrapper
      output = func(self, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  .venv/lib/python3.12/site-packages/transformers/utils/output_capturing.py:252: in wrapper
      outputs = func(self, *args, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  .venv/lib/python3.12/site-packages/transformers/models/deepseek_v3/modeling_deepseek_v3.py:612: in forward
      hidden_states = decoder_layer(
  .venv/lib/python3.12/site-packages/transformers/modeling_layers.py:93: in __call__
      return super().__call__(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  .venv/lib/python3.12/site-packages/torch/nn/modules/module.py:1778: in _wrapped_call_impl
      return self._call_impl(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  .venv/lib/python3.12/site-packages/torch/nn/modules/module.py:1789: in _call_impl
      return forward_call(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  .venv/lib/python3.12/site-packages/transformers/models/deepseek_v3/modeling_deepseek_v3.py:505: in forward
      hidden_states, _ = self.self_attn(
  .venv/lib/python3.12/site-packages/torch/nn/modules/module.py:1778: in _wrapped_call_impl
      return self._call_impl(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  .venv/lib/python3.12/site-packages/torch/nn/modules/module.py:1789: in _call_impl
      return forward_call(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  .venv/lib/python3.12/site-packages/transformers/models/deepseek_v3/modeling_deepseek_v3.py:458: in forward
      attn_output, attn_weights = attention_interface(
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  module = DeepseekV3Attention(
    (q_a_proj): Linear(in_features=8, out_features=1536, bias=False)
    (q_a_layernorm): DeepseekV3R...inear(in_features=512, out_features=1024, bias=False)
    (o_proj): Linear(in_features=512, out_features=8, bias=False)
  )
  query = tensor([[[[-1.5913, -0.4258,  1.0821,  ...,  0.1116,  0.6965, -3.0395],
            [ 2.3237, -2.0076, -0.0970,  ...,  0...,  1.1972,  1.9291],
            [ 0.2909,  0.7495, -0.7375,  ..., -1.0606, -1.0525, -1.6437]]]],
         device='cuda:0')
  key = tensor([[[[-1.1536, -0.1775,  1.1062,  ...,  0.0365,  0.0376, -0.0673],
            [ 0.7039, -0.4781, -0.4133,  ...,  0...,  0.0294, -0.2505],
            [ 0.2141,  1.8040,  0.5730,  ..., -0.2976,  0.0958, -0.0030]]]],
         device='cuda:0')
  value = tensor([[[[-2.5419e-01, -5.3610e-01,  3.8285e-03,  ..., -5.5188e-01,
              7.0597e-01,  7.8717e-01],
            [...   [-5.2057e-01,  8.4638e-01, -4.1962e-01,  ...,  5.3161e-01,
             -1.0024e+00, -1.2606e-01]]]], device='cuda:0')
  attention_mask = None, dropout = 0.0, scaling = 0.07216878364870322
  is_causal = True
  kwargs = {'position_ids': tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]],
         device='cuda:0'), 'use_cache': True, 'num_items_in_batch': 24}
  
      def sdpa_attention_forward(
          module: torch.nn.Module,
          query: torch.Tensor,
          key: torch.Tensor,
          value: torch.Tensor,
          attention_mask: torch.Tensor | None,
          dropout: float = 0.0,
          scaling: float | None = None,
          is_causal: bool | None = None,
          **kwargs,
      ) -> tuple[torch.Tensor, None]:
          if kwargs.get("output_attentions", False):
              logger.warning_once(
                  "`sdpa` attention does not support `output_attentions=True`."
                  " Please set your attention to `eager` if you want any of these features."
              )
          sdpa_kwargs = {}
          if hasattr(module, "num_key_value_groups") and module.num_key_value_groups > 1:
              if not use_gqa_in_sdpa(attention_mask, key, value):
                  key = repeat_kv(key, module.num_key_value_groups)
                  value = repeat_kv(value, module.num_key_value_groups)
              else:
                  sdpa_kwargs = {"enable_gqa": True}
      
          # Instead of relying on the value set in the module directly, we use the is_causal passed in kwargs if it is presented
          is_causal = is_causal if is_causal is not None else getattr(module, "is_causal", True)
      
          # SDPA's Flash Attention (and cuDNN) kernels rely on the `is_causal` flag. However, there are certain conditions:
          # - Not in decoding phase (otherwise we want full attention on the single query token)
          # - Attention mask is not to be provided (even if it is a causal pattern)
          # - Internally, we marked this as compatible with causal, i.e. it is a decoder attention type
          #
          # Quirks on the conditionals:
          # - We avoid inline passing this to the SDPA function directly to support both torch.compile's dynamic shapes and
          #   full graph options. Otherwise, dynamic shapes are prevented from compiling.
          # - It is important to check first for the shape, otherwise compile will fail with
          #   `argument 'is_causal' must be bool, not SymBool`.
          is_causal = query.shape[2] > 1 and attention_mask is None and is_causal
      
          # Shapes (e.g. query.shape[2]) are tensors during jit tracing, resulting in `is_causal` being a tensor.
          # We convert it to a bool for the SDPA kernel that only accepts bools.
          if torch.jit.is_tracing() and isinstance(is_causal, torch.Tensor):
              is_causal = is_causal.item()
      
          # When `is_causal = False` and the `attention_mask` is not of boolean type, the Ascend NPU's SDPA interface cannot utilize the FlashAttentionScore operator，
          # and falls back to small-operator concatenation. To invoke the FlashAttentionScore, the attention_mask must be converted to boolean type.
          # This adaptation ensures the `attention_mask` meets the requirement for using FlashAttentionScore.
          if _is_torch_npu_available:
              if attention_mask is not None and attention_mask.dtype != torch.bool:
                  # Convert to boolean type, making sdpa to force call FlashAttentionScore to improve performance.
                  attention_mask = torch.logical_not(attention_mask.bool()).to(query.device)
      
  >       attn_output = torch.nn.functional.scaled_dot_product_attention(
              query,
              key,
              value,
              attn_mask=attention_mask,
              dropout_p=dropout,
              scale=scaling,
              is_causal=is_causal,
              **sdpa_kwargs,
          )
  E       RuntimeError: The size of tensor a (4) must match the size of tensor b (8) at non-singleton dimension 1
  
  .venv/lib/python3.12/site-packages/transformers/integrations/sdpa_attention.py:93: RuntimeError
```

--- a/scripts/generate_tiny_models/for_causal_lm/deepseek_v3_for_causal_lm.py
+++ b/scripts/generate_tiny_models/for_causal_lm/deepseek_v3_for_causal_lm.py
@@ -34,7 +34,12 @@ generation_config = GenerationConfig.from_pretrained(MODEL_ID)
 config = DeepseekV3Config(
     vocab_size=len(tokenizer.vocab),
     hidden_size=8,
-    num_attention_heads=4,
+    # DeepSeek-V3 uses MLA, not GQA: kv_b_proj always reconstructs K/V with the full
+    # num_attention_heads, so num_key_value_heads must equal num_attention_heads (i.e.
+    # num_key_value_groups == 1). A mismatch (e.g. 4/2) is silently ignored by the KV
+    # projection but makes SDPA repeat_kv the already-full K/V heads, crashing with a
+    # head-count mismatch (see transformers#46960).
+    num_attention_heads=2,
     num_key_value_heads=2,
     num_hidden_layers=2,
     intermediate_size=32,

--- a/scripts/generate_tiny_models/for_causal_lm/deepseek_v3_for_causal_lm_0528.py
+++ b/scripts/generate_tiny_models/for_causal_lm/deepseek_v3_for_causal_lm_0528.py
@@ -36,7 +36,12 @@ generation_config = GenerationConfig.from_pretrained(MODEL_ID)
 config = DeepseekV3Config(
     vocab_size=len(tokenizer.vocab),
     hidden_size=8,
-    num_attention_heads=4,
+    # DeepSeek-V3 uses MLA, not GQA: kv_b_proj always reconstructs K/V with the full
+    # num_attention_heads, so num_key_value_heads must equal num_attention_heads (i.e.
+    # num_key_value_groups == 1). A mismatch (e.g. 4/2) is silently ignored by the KV
+    # projection but makes SDPA repeat_kv the already-full K/V heads, crashing with a
+    # head-count mismatch (see transformers#46960).
+    num_attention_heads=2,
     num_key_value_heads=2,
     num_hidden_layers=2,
     intermediate_size=32,

